hazelcast/src/main/java/com/hazelcast/cache/impl/merge/policy/CacheMergePolicyProvider.java

```
@@ -0,0 +1,75 @@
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.util.ConcurrencyUtil;
+import com.hazelcast.util.ConstructorFunction;
+import com.hazelcast.util.ExceptionUtil;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.hazelcast.nio.ClassLoaderUtil.newInstance;
+
+/**
+ * A provider for {@link com.hazelcast.cache.CacheMergePolicy} instances.
+ */
+public final class CacheMergePolicyProvider {
+
+    private final ConcurrentMap<String, CacheMergePolicy> mergePolicyMap;
+    private final NodeEngine nodeEngine;
+
+    private final ConstructorFunction<String, CacheMergePolicy> policyConstructorFunction
+            = new ConstructorFunction<String, CacheMergePolicy>() {
+        @Override
+        public CacheMergePolicy createNew(String className) {
+            try {
+                return newInstance(nodeEngine.getConfigClassLoader(), className);
+            } catch (Exception e) {
+                nodeEngine.getLogger(getClass()).severe(e);
+                throw ExceptionUtil.rethrow(e);
+            }
+        }
+    };
+
+    public CacheMergePolicyProvider(NodeEngine nodeEngine) {
+        this.nodeEngine = nodeEngine;
+        mergePolicyMap = new ConcurrentHashMap<String, CacheMergePolicy>();
+        addOutOfBoxPolicies();
+    }
+
+    private void addOutOfBoxPolicies() {
+        for (BuiltInCacheMergePolicies mergePolicy : BuiltInCacheMergePolicies.values()) {
+            final CacheMergePolicy cacheMergePolicy = mergePolicy.newInstance();
+            // Register `CacheMergePolicy` by its constant
+            mergePolicyMap.put(mergePolicy.name(), cacheMergePolicy);
+            // Register `CacheMergePolicy` by its name
+            mergePolicyMap.put(mergePolicy.getName(), cacheMergePolicy);
+        }
+    }
+
+    public CacheMergePolicy getMergePolicy(String className) {
+        if (className == null) {
+            throw new NullPointerException("Class name is mandatory!");
```

Yes, and I expect fail fast here or null checking should be done somewhere else.